### PR TITLE
Add Proof Set Ownership Change Logic to PandoraService

### PR DESCRIPTION
**Pull request by Hany-Almnaem**
## Overview

This PR introduces robust support for proof set ownership changes in the `PandoraService` contract, as well as tests validating this new behavior. This feature enables secure key rotation and improves the protocol's flexibility for storage providers, while maintaining registry and payment rail consistency. This work addresses issue [#15 ]

---

## Contract Changes

* **New Event:**

  * `ProofSetOwnershipChanged(uint256 indexed proofSetId, address indexed oldOwner, address indexed newOwner)` is emitted whenever proof set ownership is transferred, improving transparency for off-chain tracking.

* **Updated Logic in `ownerChanged`:**

  * Validates proof set existence and correct old owner.
  * Ensures new owner is not zero address and not already a registered provider (avoiding registry corruption).
  * Updates the internal payee and provider registry mappings, transferring the provider's identity and approval to the new owner.
  * **Note:** As I note the Payments contract currently does **not** support updating the payee of an existing rail. This update only changes internal state within `PandoraService`.

    * The contract emits an event and updates storage, but future payments will still be sent to the old owner until Payments is extended.
  * Internal comments and code are provided to guide future Payments contract upgrades.

---

## Test Changes

* **Added/Refactored Tests in `PandoraService.t.sol`:**

  * Extensive unit and integration tests for the new ownership transfer logic, covering:

    * Happy-path transfer for registered providers (ensuring registry transfer and event emission).
    * Multiple edge cases and revert conditions, including:

      * Zero address as new owner.
      * Attempting transfer to an already registered provider.
      * Non-existent proof sets.
      * Unauthorized callers.
      * Transfers with extra/large data payloads.
      * Key rotation scenarios to unregistered addresses.
    * Multiple sequential transfers, client relationship preservation, and full registry updates.

---

## Next Steps / Follow-Up

* **Payments Contract Integration Needed:**
  The actual on-chain payment flow (i.e., funds sent to the new owner after a transfer) is not yet handled by this PR.

  * If there is a plan to fully enable ownership transfers, we need to extend the Payments contract (e.g., with an `updateRailPayee` function) and integrate it into the `ownerChanged` flow.

---

## Dependency & Build Improvements

* As done in PDP, future PRs should remove all `npm` usage from the build process, to unify dependency management under git submodules and Foundry.

  * I am happy to take on these next steps and will propose follow-up PRs for each one.

---

**Please review and provide feedback.** If the overall approach and integration look good, I will proceed to:

* Remove all `npm` usage from the build process.
* Upgrade the Payments contract to support dynamic payee updates.

---

### CODE REVIEW

File: service_contracts/src/PandoraService.sol [hunk 1]
```diff
@@ -645,6 +646,66 @@ contract PandoraService is PDPListener, IArbiter, Initializable, UUPSUpgradeable
...
+    ) external override onlyPDPVerifier {
+        // Verify the proof set exists and validate the old owner
+        ProofSetInfo storage info = proofSetInfo[proofSetId];
+        require(info.railId != 0, "Proof set not registered with payment system");
+        require(info.payee == oldOwner, "Old owner mismatch");
+        require(newOwner != address(0), "New owner cannot be zero address");
+
+        // Extra safety, Don't allow to overwrite another provider's registry accidentally
+        require(
+            providerToId[newOwner] == 0 && !approvedProvidersMap[newOwner],
```
Comment (ZenGround0) [inline service_contracts/src/PandoraService.sol:671]: This approach breaks reasonable assumptions if an owner has more than one proofset.  In that case other proofsets will continue to reference the old owner who will no longer be an approved provider.

This change cascades proofset ownership changes into actual modification of the existing SP registration map.  And therefore it has to check cases like this in order to keep the registration map updated.  

Instead what we should do is completely decouple registration of SP addresses from ownership changes.  The correct way to do this is:
1. remove this require
2. remove the `// Update provider registry mappings if the old owner was a registered provider` block of code entirely
3. remove event emission
4. add a check that the new owner is actually registered i.e. `require(approvedProvidersMap[newOwner])`

To make the usage pattern clear the way this will work in practice looks like this
1. Provider wants to rotate their keys
2. They register and approve a new provider key, get a completely new provider id, and mapping that id to the same pdp and retrieval urls
3. They then call change proofset owner on pdp which calls into here
4. Finally if they want to they can remove the old provider with a call to `removeServiceProvider`

This all hinges on the fact that it is ok to have multiple providers ids pointing to the same metadata.

**Comment by ZenGround0** [comment]
FYI @rjan90 @aarshkshah1992, please advise if you disagree with the simple stop gap approach for the time being.

**Comment by ZenGround0** [comment]
@Hany-Almnaem here's the plan to unblock you 
1. For this PR just implement that ownership change as described in my [review comment](https://github.com/FilOzone/filecoin-services/pull/65#discussion_r2178344578)
2. Introduce a beneficiary address for paying out rails: #68.  Note that there is no plan to change beneficiaries of a proofset.  The responsibility is on the provider to setup their beneficiary as a multisig if they want better key management.

**Comment by Hany-Almnaem** [comment]
Thanks, @ZenGround0, for your detailed feedback.

I understand the reasoning behind decoupling ownership and beneficiary logic. I’ll update this PR to:

    Remove any updates to the provider registry and related mappings in ownerChanged.

    Remove the requirement and event logic related to the provider registry.

    Add a check to ensure the new owner is an already registered provider (require(approvedProvidersMap[newOwner])).

    Update the tests to match this streamlined behavior.

After this, I’ll be happy to help with the beneficiary address issue (#68)

Will push changes shortly. Please let me know if there’s anything I missed!

**Comment by Hany-Almnaem** [comment]
Hi @ZenGround0,

Quick update:
Both the PandoraService implementation and the ownerChanged test suite have been fully refactored to follow the latest feedback:

    Implementation:

        Ownership changes are now fully decoupled from the provider registry, the contract only updates the local payee and emits the event, with no mutation of provider mappings.

    Tests:

        All ownerChanged tests now validate only local state changes, registry invariants, correct event emission, and all relevant edge/negative cases.

        Removed any checks or legacy tests expecting provider registry mutation on ownership change.

Please let me know if you spot anything that needs further adjustment.

Thanks again for your guidance and review

**Comment by rjan90** [comment]
Hey @Hany-Almnaem! Just wanted to let you know that we have not lost track of this PR, and that @ZenGround0 will give it another review today - and get it merged if everything looks good!

File: service_contracts/src/PandoraService.sol [hunk 2]
```diff
@@ -645,6 +646,66 @@ contract PandoraService is PDPListener, IArbiter, Initializable, UUPSUpgradeable
...
+     */
+    function ownerChanged(
+        uint256 proofSetId,
+        address oldOwner,
+        address newOwner,
+        bytes calldata extraData
+    ) external override onlyPDPVerifier {
+        // Verify the proof set exists and validate the old owner
+        ProofSetInfo storage info = proofSetInfo[proofSetId];
+        require(info.railId != 0, "Proof set not registered with payment system");
```
[the remark this answers is not in the document]
Comment (Hany-Almnaem) [reply service_contracts/src/PandoraService.sol:665]: Thanks for following up on this @ZenGround0 !
So when I originally added that check:

' `require(info.railId != 0, "Proof set not registered with payment system"); `'

My thinking was basically to make sure `'ownerChanged'` only got called for proof sets that actually had an active payment rail set up. I was treating it as a safeguard against running unnecessary operations or ending up with weird internal state issues. At the time, I was viewing this ownership hook as something that should only care about proof sets that were already plugged into the payment system.
But you're absolutely right about the coupling issue. After thinking about it more (and thanks to your feedback), I realized I was being a little bit restrictive. Tying ownership logic directly to payment registration like that really does limit flexibility in ways that could bite us later.

What if we need to track ownership changes for proof sets that aren't connected to payment rails yet? Or for other scenarios like auditing workflows or cases where onboarding happens in a different order? The tight coupling would make all of that unnecessarily complicated.
So yeah, removing that check was the right call.

That said, I do think there might be something interesting in the general idea of having more sophisticated validation around payment-linked operations - maybe as an optional layer or different hook entirely. Not sure if that's something worth exploring down the line, but I'd be curious to hear thoughts on it if you think there's potential value there.
